### DISCUSSION TITLE
Only sending audio data every other frame?

### DISCUSSION THREAD

**Discussion by kaspernyhus** [hathach/tinyusb#1336] [Q&A]
Hi,
I'm trying to get USB Audio working on the esp32s3 with the newest version of tinyusb. Eventually I would like to make an I2S to USB Audio bridge. I have data coming through, but sniffing on the traffic I can see that only every second frame has data included. I'm using the "audio_test" example.

Has anybody here experimented with the USB Audio device implementation? Or maybe even a I2S to USB Audio implementation?

Thanks,
Kasper Nyhus

![Screenshot from 2022-02-22 09-41-15](https://user-images.githubusercontent.com/46648238/155096031-5098e5bb-3d42-488f-aab7-dcbfa7512d97.png)
![Screenshot from 2022-02-22 09-45-55](https://user-images.githubusercontent.com/46648238/155096039-62b21245-0480-4dcf-8292-27688fc0f6fc.png)
![Screenshot from 2022-02-22 09-51-03](https://user-images.githubusercontent.com/46648238/155096454-2872fe9c-4270-4e1c-8736-d9ba70d6a424.png)

**Comment by kaspernyhus** [comment]
Follow up:
I uploaded a similar example to the Raspberry Pi Pico, and here the data is transmitted in every frame as expected! Interesting.
I'm very interested if other people have tried to implement USB Audio on the esp32

Thanks,
Kasper Nyhus

**Comment by kaspernyhus** [comment]
Every other call to osal_queue_receive() inside tud_task() takes more than 1000us, which makes it miss the 1ms usb full speed frame?


    start = (uint32_t)esp_timer_get_time();

    if ( !osal_queue_receive(_usbd_q, &event) ) return;

    end = (uint32_t)esp_timer_get_time();
    ESP_LOGI("tud_task queue recv", "Timer %d us",end-start);



I (331667) tud_task queue recv: Timer 1268 us
I (331677) tud_task queue recv: Timer 233 us
I (331677) tud_task queue recv: Timer 1287 us
I (331687) tud_task queue recv: Timer 251 us
I (331687) tud_task queue recv: Timer 1305 us
I (331697) tud_task queue recv: Timer 270 us
I (331697) tud_task queue recv: Timer 1322 us
I (331707) tud_task queue recv: Timer 287 us
I (331707) tud_task queue recv: Timer 1341 us
I (331717) tud_task queue recv: Timer 305 us
I (331717) tud_task queue recv: Timer 1359 us
I (331727) tud_task queue recv: Timer 324 us
I (331727) tud_task queue recv: Timer 1378 us
I (331737) tud_task queue recv: Timer 342 us
I (331737) tud_task queue recv: Timer 1395 us
I (331747) tud_task queue recv: Timer 359 us
I (331747) tud_task queue recv: Timer 1413 us
I (331757) tud_task queue recv: Timer 377 us
I (331757) tud_task queue recv: Timer 1432 us
I (331767) tud_task queue recv: Timer 396 us
I (331767) tud_task queue recv: Timer 1450 us
I (331777) tud_task queue recv: Timer 414 us
I (331777) tud_task queue recv: Timer 1468 us

**Reply by kaspernyhus** [reply]
seems like the esp32 is generating interrupts at alternately ~30us and ~1970us intervals. These interrupts trigger _dcd_int_handler() which makes it impossible for the tinyusb stack to meet the 1ms frame every second time.

**Comment by HiFiPhile** [comment]
Hi Kasper,

I'm busy with my works at the moment so I didn't pay much attention here.

I don't have much experience with esp32, in fact I just started my first project on esp32c3 last month.

I still prefer "old school" MCUs, although more code need to write, robustness is much controllable.

**Comment by siddhartha993549** [comment]
Hi,

Could anyone please help me to understand that What exactly is audio test example supposed to do?
